docs/style-guide/enso.md

```diff
@@ -0,0 +1,138 @@
+---
+layout: style-guide
+title: Enso Style Guide
+category: style-guide
+tags: [style-guide]
+order: 7
+---
+
+# Enso Style Guide
+
+This style guide is aimed at developers writing Enso Libraries Code in the Enso
+language. It is not aimed at users writing Enso code via the graphical editor.
+
+# Encapsulation
+
+Encapsulation allows developers to present a consistent interface that is
+independent of its internal implementation. It is the foundation of reuseable
+"pieces" of code and funcationality. It allows other users (both internal and
+external) effectivley re-use components of your code.
+
+# Types of Code Components in the Enso Langauge
+
+In order to be able to talk about how we encapsulate components of code we first
+need to define what those components are. Enso has four levels of code
+components that can be built and shared with other developers. They are:
+
+- Libaries
+- Modules
+- Types
+- Methods
+
+TODO - defintion of each of these.
+
+# Public verus Private
+
+Now we have defined our components we need to look at how we define our public
+APIs for each of these and how we hide our private immplementations.
+
+## Public
+
+By default the Enso langauge is a very public one. If you don't use any of the
+private keywords and/or conventions described belown then everything is public.
+This is bad for encapsulation as other users and developers trying to re-use the
+code you wrote won't have a clean view of your code's API.
+
+## Types of Private
+
+There are a number of different ways to mark things Private in Enso
+
+### Hidden from the Graphical Editor Private
+
+This is acheived with the special private: true keyword cotained in a comment block above a
+type, constructor or method
+
+e.g.
+
+```
+## ---
+   private: true
+   ---
+type Random_Generator
+```
+
+or
+
+```
+## ---
+   private: true
+   ---
+Value (random_instance_holder:RandomInstanceHolder)
+```
+
+or
+
+```
+## ---
+   private: true
+   ---
+set_seed self seed = self.random_instance_holder.setSeed seed
+```
+
+This prevents a type, construstor ot method being used from the Component
+Browser inside of the graphical editor.
+
+For other Enso code users it does not prevent its use and these are part of your
+components public API.
+
+### Library Private
+
+This is a langauge keyword `private` that can be used on modules, types, methods
+and constructors.
+
+e.g.
+
+TODO Examples.
+
+This keyword makes whatever it marks private at the library component level and
+so is useful in defining the public and private parts of an Enso library. It is
+enforced by the compliler, so has a strong guarentee of enforcing the Public and
+Private parts of a library.
+
+However for other developers working inside of the same library it does not
+prevent their use and these components are still part of your internal public
+API.
+
+### Module Private
+
+At the time of writing, the above is all the support the langauge has built-in
```

Comment:
- Should this be an API design guide, then this kind of privacy unimportant.
- There already is `private` module "privacy" (important from API perspective)
- "Module Private" vs. "private module" will very likely form a confusing terminology pair